clang/lib/AST/ExprConstant.cpp

```diff
@@ -7304,6 +7382,21 @@ class BufferToAPValueConverter {
     return ArrayValue;
   }
 
+  std::optional<APValue> visit(const VectorType *Ty, CharUnits Offset) {
+    SmallVector<uint8_t, 8> Bytes;
+    if (!Buffer.readObject(Offset, Info.Ctx.getTypeSizeInChars(Ty), Bytes))
+      return std::nullopt;
```

Comment:
... and here we'll need to not reject if the padding bits are uninitialized.

As a testcase, we should ensure that (for example) this:
```c++
struct A {
  long double a, b, c, d;
};
```
can round-trip through a vector of 4 `long double`s.

Reply:
At least, assuming that there is padding in the same places in both. Maybe the vector doesn't actually have padding, but the struct does, so the round-trip *should* fail?

Reply:
Hrm, this is indeed awkward: I obviously can't load the entire f80 vector at once, but loading individual elements at a time would not only force me to duplicate much of the code again, but would also make bool vectors break again without once again adding special handling (since we'd have to load individual bytes at a time again).

Can we simply assume the uninitialized padding bits are always zero? This test in `CodeGen/const-init.c` (the case `b`):  
https://github.com/llvm/llvm-project/blob/8768741800aae37a825864e2ee782484ed073ce9/clang/test/CodeGen/const-init.c#L138-L148  
which uses `CK_BitCast` seems to make that assumption, while a variety of the existing tests in `SemaCXX/constexpr-builtin-bit-cast.cpp`:  
https://github.com/llvm/llvm-project/blob/8768741800aae37a825864e2ee782484ed073ce9/clang/test/SemaCXX/constexpr-builtin-bit-cast.cpp#L332-L351  
assume that bit-casting padding space should return uninitialized values (or fail when the target type isn't allowed to be uninitialized in a constexpr context).

Is the `CK_BitCast` assumption simply wrong, or do these two different ways of triggering a bit cast really have such different semantics? I can see a few places in the actual LLVM constant folding code where vector bit casts containing uninitialized elements are, in fact, initialized to zero, but I feel like that makes sense there since it doesn't have to adhere to the rules of `constexpr`. Should I rewrite all the bit cast conversion code in `APValueToBufferConverter`/`BufferToAPValueConverter` again while taking uninitialized padding bytes into account, and redirect the existing `CK_BitCast` code to point to that?

Reply:
The object representation indeed does not seem to initialize the padding bits: https://godbolt.org/z/szWenbE8c

- In Clang, the padding bytes contain non-zero values for both vectors and structs.
- In Clang msan, only the branch on uninitialized padding in the struct is caught. (I think this is an msan bug; it presumably isn't taking into account that a vector type might contain padding.)
- In GCC, the padding bytes in the vector are zeroed but the padding bytes in the struct are not. (Though that could be an accident of implementation rather than intent.)

Casts between vector types don't seem to initialize padding bits either: https://godbolt.org/z/Gex93x3xx -- so the test in const-init.c seems to be wrong.

> Should I rewrite all the bit cast conversion code in `APValueToBufferConverter`/`BufferToAPValueConverter` again while taking uninitialized padding bytes into account, and redirect the existing `CK_BitCast` code to point to that?

:( Yeah, I guess so. Sorry.

Reply:
Hm, maybe we need to avoid generating an LLVM `bit_cast` here and go through memory instead?

I think it's best to handle that in a separate PR.